plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSig/Stage0.hs

```diff
@@ -0,0 +1,247 @@
+    ||]
+
+pkListEq :: Q (TExp ([PubKey] -> [PubKey] -> Bool))
+pkListEq = [||
+
+    let pkListEq' :: [PubKey] -> [PubKey] -> Bool
+        pkListEq' [] [] = True
+        pkListEq' (k:ks) (k':ks') = $$(P.and) ($$(Validation.eqPubKey) k k') (pkListEq' ks ks')
+        pkListEq' _ _ = False
+    
+    in pkListEq'
+
+    ||]
+
+-- | Check whether a proposed 'Payment' is valid given the total
+--   amount of funds currently locked in the contract.
+isValidProposal :: Q (TExp (Value -> Payment -> Bool))
+isValidProposal = [||
+
+        let isValidProposal' :: Value -> Payment -> Bool
+            isValidProposal' vl (Payment amt _ _) = $$(Value.leq) amt vl
+        in isValidProposal'
+
+    ||]
+
+-- | Check whether a proposed 'Payment' has expired.
+proposalExpired :: Q (TExp (PendingTx -> Payment -> Bool))
+proposalExpired = [|| 
+    \(PendingTx _ _ _ _ _ rng _ _) (Payment _ _ ddl) -> 
+        $$(Slot.before) ddl rng
+    ||]
+
+-- | Check whether enough signatories (represented as a list of public keys)
+--   have signed a proposed payment.
+proposalAccepted :: Q (TExp (Params -> [PubKey] -> Bool))
+proposalAccepted = [||
+        let proposalAccepted' :: Params -> [PubKey] -> Bool
+            proposalAccepted' (Params signatories numReq) pks = 
+                let numSigned = $$(P.length) ($$(P.filter) (\pk -> $$containsPk pk pks) signatories)
+                in $$(P.geq) numSigned numReq
+        in proposalAccepted'
+    ||]
+
+-- | @valuePreserved v p@ is true if the pending transaction @p@ pays the amount
+--   @v@ to a single pay-to-script output at this script's address.
+valuePreserved :: Q (TExp (Value -> PendingTx -> Bool))
+valuePreserved = [||
+
+        let valuePreserved' :: Value -> PendingTx -> Bool
+            valuePreserved' vl ptx = 
+                let ownHash = $$(Validation.ownHash) ptx
+                    numOutputs = $$(P.length) ($$(Validation.scriptOutputsAt) ownHash ptx)
+                    valueLocked = $$(Validation.valueLockedBy) ptx ownHash
+                in $$(P.and) ($$(P.eq) 1 numOutputs) ($$(Value.eq) valueLocked vl)
+        in valuePreserved'
+
+    ||]
+
+-- | @valuePaid pm ptx@ is true if the pending transaction @ptx@ pays
+--   the amount specified in @pm@ to the public key address specified in @pm@
+valuePaid :: Q (TExp (Payment -> PendingTx -> Bool))
+valuePaid = [||
+
+        let valuePaid' :: Payment -> PendingTx -> Bool
+            valuePaid' (Payment vl pk _) ptx = $$(Value.eq) vl ($$(Validation.valuePaidTo) ptx pk)
+        in valuePaid'
+    ||]
+
+-- | Equality of 'Payment' values
+paymentEq :: Q (TExp (Payment -> Payment -> Bool))
+paymentEq = [||
+
+        let paymentEq' :: Payment -> Payment -> Bool
+            paymentEq' (Payment vl pk sl) (Payment vl' pk' sl') = 
+                $$(P.and) ($$(P.and) ($$(Value.eq) vl vl') ($$(Validation.eqPubKey) pk pk')) ($$(Slot.eq) sl sl')
+        in paymentEq'
+
+    ||]
+
+-- | Equality of 'State' values
```

Comment:
Perhaps stupid question, but why do we need equality of `State` values in `StateMachine`?

Reply:
In the validator script we compute the next state, and then we need to compare that with the next state that was given in the redeemer: https://github.com/input-output-hk/plutus/blob/master/plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs#L53

So ultimately it is to make sure that the next data script contains the correct state.

Reply:
Would there be any sense in not storing the current state in the redeemer and store only inputs? State can be huge, equality checks can be expensive. And less code -- less space for bugs. State can be reconstructed from inputs and the initial state, right?

Reply:
Yeah that's a good idea, but applying the step function repeatedly results in higher gas cost so I think we need to decide on a case-by-case basis which option is cheaper.

Reply:
> applying the step function repeatedly 

Do we need to do this on-chain? I mean, the validator knows its state and can apply the step function directly to it, no need to recompute the state from grounds up. Validation of the entire blockchain happens off-chain. What am I missing?

Reply:
We currently need the redeemer script and the data script of the next transaction to be identical, due to the "validating the next data script" problem (if they're identical then we can check that the next data script is correct because its hash should be the same as that of the redeemer)